File: kd_tree.cpp
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <time.h>
#include <bits/stdc++.h>
#define pq  priority_queue
#define MAX_DIM 3
using namespace std;
struct kd_node_t{
    double x[MAX_DIM];
    double dis,rad; // sort by a closest then farthest
    bool operator<(const struct kd_node_t& other) const{return dis < other.dis;} // logic
    struct kd_node_t *left, *right;
};
 
    inline double
dist(struct kd_node_t *a, struct kd_node_t *b, int dim)
{
    double t, d = 0;
    while (dim--) {
        t = a->x[dim] - b->x[dim];
        d += t * t;
    }
    return d;
}
inline void swap(struct kd_node_t *x, struct kd_node_t *y) {
    double tmp[MAX_DIM];
    memcpy(tmp,  x->x, sizeof(tmp));
    memcpy(x->x, y->x, sizeof(tmp));
    memcpy(y->x, tmp,  sizeof(tmp));
}
 
 
/* see quickselect method */
    struct kd_node_t*
find_median(struct kd_node_t *start, struct kd_node_t *end, int idx)
{
    if (end <= start) return NULL;
    if (end == start + 1)
        return start;
 
    struct kd_node_t *p, *store, *md = start + (end - start) / 2;
    double pivot;
    while (1) {
        pivot = md->x[idx];
 
        swap(md, end - 1);
        for (store = p = start; p < end; p++) {
            if (p->x[idx] < pivot) {
                if (p != store)
                    swap(p, store);
                store++;
            }
        }
        swap(store, end - 1);
 
        /* median has duplicate values */
        if (store->x[idx] == md->x[idx])
            return md;
 
        if (store > md) end = store;
        else        start = store;
    }
}
 
    struct kd_node_t*
make_tree(struct kd_node_t *t, int len, int i, int dim)
{
    struct kd_node_t *n;
 
    if (!len) return 0;
 
    if ((n = find_median(t, t + len, i))) {
        i = (i + 1) % dim;
        n->left  = make_tree(t, n - t, i, dim);
        n->right = make_tree(n + 1, t + len - (n + 1), i, dim);
    }
    return n;
}
 
/* global variable, so sue me */
int visited;
 
set<kd_node_t> bruter;
void nearest(struct kd_node_t *root, struct kd_node_t *nd, int i, int dim,
        struct kd_node_t **best, double *best_dist,int size)
{
    double d, dx, dx2;
 
    if (!root) return;
    d = dist(root, nd, dim);
    dx = root->x[i] - nd->x[i];
    dx2 = dx * dx;
 
    visited ++;

    root->dis = d;    // Putting inside pq with priority distance
    root->rad = dx2;
    bruter.insert(*root);
    while(bruter.size()>size){
    set<kd_node_t>::iterator it = bruter.end();it--;bruter.erase(it);};
 
    if (!*best || d < *best_dist) {
        *best_dist = d;
        *best = root;
    }
  
    if (++i >= dim) i = 0;
    nearest(dx > 0 ? root->left : root->right, nd, i, dim, best, best_dist,size);
    set<kd_node_t>::iterator it = bruter.end();it--;
    if((bruter.size() < size || it->dis >= dx2))
    nearest(dx > 0 ? root->right : root->left, nd, i, dim, best, best_dist,size);
}
 
#define N 1000000
#define rand1() (rand() / (double)RAND_MAX)
#define rand_pt(v) { v.x[0] = rand1(); v.x[1] = rand1(); v.x[2] = rand1(); }
struct kd_node_t wp[] = {
        {{2, 3},123,0}, {{5, 4},123,0}, {{9, 6},123,0}, {{3, 7},123,0},
         {{4, 7},123,0}, {{8, 1},123,0}, {{7, 2},123,0}, {{6, 2},123,0},
         {{14, 17},123,0},{{2, 2},123,0},{{6, 6},123,0},{{11, 2},123,0},
         {{7, 4},123,0},{{9, 1},123,0},{{5, 1},123,0},{{7, 1},123,0}    };
struct kd_node_t testNode = {{9, 2},123,0};
struct kd_node_t *root, *found, *million;



void brute( struct kd_node_t *nd)
{
   for(int i=0;i<sizeof(wp) / sizeof(wp[1]);i++)
   {
    wp[i].dis=dist(&wp[i],nd,2);
    bruter.insert(wp[i]);
   }
   cout << "\nThis is the test node and these are the results of brute force \n " << testNode.x[0] << " " << testNode.x[1] << "\n\n\n";
   set<kd_node_t>::iterator it;
   for(int i=0;i<10;i++)
   {

    it = bruter.begin();
    cout << it->x[0] <<"  " << it->x[1] <<"  " << it->dis << "\n";
    bruter.erase(it);
   }

}

set<kd_node_t> asd;
void brute2( struct kd_node_t *nd,int size)
{
   for(int i=0;i<N;i++)
   {
    million[i].dis=dist(&million[i],nd,3);
    asd.insert(million[i]);
   }
   cout << "\nThis is the test node " << testNode.x[0] << " " << testNode.x[1] << " " << testNode.x[2]<<"\n and these are the results of brute force \n "<< "\n";
   set<kd_node_t>::iterator it;
   for(int i=0;i<size;i++)
   {
    it = asd.begin();
    cout << it->x[0] <<"  " << it->x[1] <<"  " << it->dis << "\n";
    asd.erase(it);
   }

}
int main(void)
{
    int i;
    double best_dist;
 
    root = make_tree(wp, sizeof(wp) / sizeof(wp[1]), 0, 2);
 
    visited = 0;
    found = 0;
    nearest(root, &testNode, 0, 2, &found, &best_dist,8);
 
    printf(">> WP tree\nsearching for (%g, %g)\n"
            "found (%g, %g) dist %g\nseen %d nodes\n\n",
            testNode.x[0], testNode.x[1],
            found->x[0], found->x[1], sqrt(best_dist), visited);
    set<kd_node_t>::iterator it;
    while(bruter.size()){
        it = bruter.begin();
        cout << it->x[0] <<"  " << it->x[1] <<"  " << it->dis << "\n";
        bruter.erase(it);
    };
    
    
    brute(&testNode);
    while(bruter.size()){
        it = bruter.begin();
        bruter.erase(it);
    };
    
million =(struct kd_node_t*) calloc(N, sizeof(struct kd_node_t));
srand(time(0));
    for (i = 0; i < N; i++) rand_pt(million[i]);
 
    root = make_tree(million, N, 0, 3);
    rand_pt(testNode);
 
    visited = 0;
    found = 0;
    int size=14;
       clock_t t1,t2;
    t1=clock();
    //code goes here
    nearest(root, &testNode, 0, 3, &found, &best_dist,size);
 
    t2=clock();
    float diff ((float)t2-(float)t1);

    printf(">> Million tree\nsearching for (%g, %g, %g)\n"
            "found (%g, %g, %g) dist %g\nseen %d nodes\n",
            testNode.x[0], testNode.x[1], testNode.x[2],
            found->x[0], found->x[1], found->x[2],
            (best_dist), visited);
     while(bruter.size()){
        it = bruter.begin();
        cout << it->x[0] <<"  " << it->x[1] <<"  " << it->x[2]  <<"  "<< it->dis << "\n";
        bruter.erase(it);
    };
       t1=clock();
    brute2(&testNode,size);
    nearest(root, &testNode, 0, 3, &found, &best_dist,size);
 
    t2=clock();
    float diff2 ((float)t2-(float)t1);
    cout<< "\nTime taken by brute force \n"<<diff2<<endl; 
    cout<< "\nTime taken by kd tree \n"<<diff<<endl;
    cout << "The ratio of time taken by k-d tree and highly optimized brute force(O(N Log N)) " << diff2/diff << "\n";
        return 0;
}

File: kd_tree.rb
require "rbtree"
class Kdtree
  include Enumerable
  class Node
    def initialize(point, left=nil, right = nil, dim = nil, distance = nil, radius = nil)
      @point = point
      @left = left
      @right = right
      @distance = distance
      @radius = radius
      @dim = dim
    end
    attr_accessor :point, :left, :right, :distance, :radius
  end
  def initialize(points,dim,i=0)
    @points = points
    @root = make_tree(points,dim,i,1)
    @dime = dim
    @best_dist = 1000000000
  end
  def dist(a,b)
    size=a.size
    sum=0
    size.times do |i|
      sum= sum + (a[i]-b[i])*(a[i]-b[i])
    end
    sum
  end
  def quickselect(a, k)
    arr = a.dup
    loop do
      pivot = arr.delete_at(rand(arr.length))
      left, right = arr.partition { |x| x < pivot }
      if k == left.length
        return pivot
      elsif k < left.length
        arr = left
      else
        k = k - left.length - 1
        arr = right
      end
    end
  end
  def make_tree(points,dim,i,index)
 #   print points,"\n"
    root=Kdtree::Node.new(points[0])
    dimension_array=[]
    size = points.size
    points.each do |point|
      dimension_array.push(point[i]) # for each j point we choose dim i
    end
    pivot = quickselect(dimension_array, (dimension_array.size-1) / 2)
 #   puts pivot
    left = []
    right = []
    qwe=0
    points.each do |point|
      left.push(point) if point[i]<pivot
      right.push(point) if point[i]>pivot
    #  puts point[i]
      if point[i]==pivot and qwe==1
    #    puts "asdlksa"
        left.push(point)
        qwe=1
      end
      if point[i]==pivot and qwe==0
    #    puts "asdlksa"
        root=Kdtree::Node.new(point)
        qwe=1
      end

    end
 #   puts root.point[i]
  #  puts "\n\n\n"
    if left.size==0
      root.left=nil
    end
    if left.size==1
      root.left = Kdtree::Node.new(left[0])
    end
    if left.size>1
      root.left = make_tree(left,dim,(i + 1) % dim,index+1)
    end
    if right.size==0
      root.right=nil
    end
    if right.size==1
      root.right=Kdtree::Node.new(right[0])
    end
    if right.size>1
       root.right = make_tree(right,dim,(i + 1) % dim,index+1)
     end

    root
  end
  def near(nearest,size)
    rbtree=MultiRBTree[]
    a=nearest(root,nearest,0,@dime,nil,100000,0,rbtree,size)
  end
  
  def nearest(root,nearest,i,dim,best,best_dist,index,rbtree,size)
    return rbtree if root==nil
    d = dist(root.point,nearest)
    dx = root.point[i]-nearest[i]
    dx2= dx * dx

    root.distance=d
    root.radius = dx2
    rbtree[d]=root.point
    while rbtree.size>size do
      rbtree.pop
    end
    rbtree=nearest(dx > 0 ? root.left : root.right, nearest, (i+1)%dim, dim, best, best_dist,index+1,rbtree,size)
    if (rbtree.size< size || rbtree.last[0]>=dx2) 
      rbtree=nearest(dx > 0 ? root.right : root.left, nearest, (i+1)%dim, dim, best, best_dist,index+1,rbtree,size)
    end
    rbtree
  end
  attr_accessor :root, :point
end

c=Kdtree.new(   [[2, 3],[5, 4], [6, 6], [6, 6],
         [4, 7], [8, 1], [7, 2], [6, 2],[4,6],
         [14, 17],[2, 2],[6, 6],[11, 2],
         [7, 4],[9, 1],[5, 1],[7, 1]],    2)
a=c.near([5, 8],8)
siz=a.size
b=[]
while a.size>0 do
      b.push(a.pop)
end
siz.times do |i|
  print b[siz-1-i][1]," ",b[siz-1-i][0],"\n"
end
a=Time.now()
10000.times do |i|
  c.near([9,2],9)
end
puts Time.now()-a


c=[]
100000.times do |i|
  a=rand(523)
  b=rand(523)
  d=rand(523)
  c.push([a,b,d])
end
asdgh=c
d=Kdtree.new(c,3)
qw=Time.now()
x=d.near([13, 234,444],5)
siz=x.size
b=[]
while x.size>0 do
      b.push(x.pop)
end
ans1=[]
ans2=[]
siz.times do |i|
  print b[siz-1-i][1],b[siz-1-i][0],"\n"
  ans1.push(b[siz-1-i][0])
end

puts t1=Time.now()-qw
def dist(a,b)
    size=a.size
    sum=0
    size.times do |i|
      sum= sum + (a[i]-b[i])*(a[i]-b[i])
    end
    sum
end
dister=[]
qw=Time.now()
c.each do |i|
    dister.push(i.push(dist([13, 234,444],i)))
end
dister=dister.sort{|a,b| a[3] <=> b[3]}
5.times do |i|
  print dister[i],"\n"
  ans2.push(dister[i][3])
end
puts t2=Time.now()-qw
print "Ratio of time taken by brute force and kd tree ", t2/t1,"\n"
print asdgh==c,"\n",ans1==ans2,"\n"
